quark/foreach.py

```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 *--
+
+import os
+import subprocess
+from argparse import ArgumentParser
+
+from quark.utils import DirectoryContext
+
+from .subproject import Subproject
+
+
+def run():
+    parser = ArgumentParser(
+        description="""
+        Evaluates an arbitrary shell command in each submodule, skipping all
+        the svn's submodules.
+        The command has access to the variables $name, $sm_path, $displaypath,
+        $sha1, $toplevel, $rev, $version_control:
+        $name is the name of the submodule;
+        $sm_path is the path of the submodule relative to the superproject;
+        $displaypath is the path of the submodule relative to the root
+        directory;
+        $version_control is the version control used by the subproject
+        (git/svn);
+        $sha1 is the commit of the subproject ( empty string if it is a
+        svn repository );
+        $rev is the revision of the subproject ( empty string if it is a
+        git repository );
+        $toplevel is the absolute path to the top-level of the
+        immediate superproject.
+        """
+    )
+    parser.add_argument(
+        "-q", "--quiet",
+        action="store_true",
+        help="Only print error messages"
+    )
+    parser.add_argument(
+        "command",
+        action="store",
+        nargs="+",
+        help="The command that will be run for every dependency"
+    )
+
+    optlist = parser.parse_args()
+
+    root, modules = Subproject.create_dependency_tree(
+        os.getcwd(), update=False
+    )
+
+    for path, module in modules.items():
+        cmd = " ".join([
+            os.path.abspath(x) if os.path.exists(x) else x
```

Comment:
I just noticed this and, if it is what it seems (absolutizing stuff that looks like paths so that it will continue to work even when the command is run in each subproject directory), it's a terrible idea:

- shell syntax (_which_ shell, by the way?) is a mess; brutally splitting on spaces hoping to get the arguments of the command the user put in is _extremely_ wishful thinking;
- even admitting you got the arguments right, deciding to absolutize paths using as heuristic whether they exist in the _current_ directory or not is a no-no. It's perfectly plausible that my script wants to operate on files that exists both in the top-level project and in the submodules; even worse, there may be non-file arguments (or even command names!) that happen to match files existing in the current directory.

So, this whole thing must go. Shell commands are _for all practical purposes_ opaque sequences of bytes¹ that shouldn't be touched with a ten foot pole, but just forwarded straight to the shell. It's enough of a pain to deal with shell syntax without having intermediate code mess with it.

---

1. Actually, Python 3 being stupid, here they are actually already decoded to _Unicode_ strings, which would break stuff if not for the kludge that is PEP 383. But I'm digressing.

Reply:
Removed in https://github.com/comelz/quark/pull/27/commits/3e51c6636c9b4f9e778022c64d9795cfdfa11ab9

I've added some tests for it in https://github.com/comelz/quark/pull/27/commits/e3864bdc6799bab4fdc7c7c33cd9094e2c77bb32